### ISSUE TITLE
Exception while parsing mmCif files

### ISSUE THREAD

**Issue by ofajardo** [biopython/biopython#778]
I get an error while parsing some mmCif files. Would it be possible to fix it?
I am using the latest version of biopython in github (1.66+), python 2.7.10, both in linux and windows.

When does the error happens?
Just doing this:
MMCIF_parser = MMCIFParser()
MMCIF_structure = MMCIF_parser.get_structure(protein_ID, pdb_file_path)

structures affected: 4Q96 and 4Q94

**Structure: 4Q96**
traceback: Traceback (most recent call last):
MMCIF_structure = MMCIF_parser.get_structure(protein_ID, pdb_file_path)
File "/homebasel/biocomp/fajardoo/python_test_modules/biopython_repo/Bio/PDB/MMCIFParser.py", line 65, in get_structure
self._build_structure(structure_id)
File "/homebasel/biocomp/fajardoo/python_test_modules/biopython_repo/Bio/PDB/MMCIFParser.py", line 181, in _build_structure
structure_builder.init_residue(resname, hetatm_flag, int_resseq, icode)
File "/homebasel/biocomp/fajardoo/python_test_modules/biopython_repo/Bio/PDB/StructureBuilder.py", line 157, in init_residue
% (resname, field, resseq, icode))
PDBConstructionException: Blank altlocs in duplicate residue ARG (' ', 55, ' ')

**Structure 4Q94**
traceback: Traceback (most recent call last):
MMCIF_structure = MMCIF_parser.get_structure(protein_ID, pdb_file_path)
File "/homebasel/biocomp/fajardoo/python_test_modules/biopython_repo/Bio/PDB/MMCIFParser.py", line 65, in get_structure
self._build_structure(structure_id)
File "/homebasel/biocomp/fajardoo/python_test_modules/biopython_repo/Bio/PDB/MMCIFParser.py", line 181, in _build_structure
structure_builder.init_residue(resname, hetatm_flag, int_resseq, icode)
File "/homebasel/biocomp/fajardoo/python_test_modules/biopython_repo/Bio/PDB/StructureBuilder.py", line 157, in init_residue
% (resname, field, resseq, icode))
PDBConstructionException: Blank altlocs in duplicate residue SER (' ', 2, ' ')

**Comment by JoaoRodrigues** [comment]
@ofajardo Thanks again!

I actually get a different error when parsing both those structures. Which version of Biopython are you using exactly? I took the structures directly from the RCSB PDB (not biological unit).

```python
/Users/joao/software/biopython/build/lib.macosx-10.11-intel-2.7/Bio/PDB/StructureBuilder.py in init_chain(self, chain_id)
     81         o chain_id - string
     82         """
---> 83         if self.model.has_id(chain_id):
     84             self.chain = self.model[chain_id]
     85             warnings.warn("WARNING: Chain %s is discontinuous at line %i."

AttributeError: 'StructureBuilder' object has no attribute 'model'
```

**Comment by ofajardo** [comment]
Hi, 

I downloaded on monday I think biopython from github as a zip. The structures I got from ftp.wwpdb.org. It is very funny, I just tested again these two lines in isolation and now I get exactly the same error as you do (see below). Then I run my program only for those two structuers and got the same. I don't know exactly why I got that strange error before, I think there is something wrong with my linux environment. Anyway, you're right with the error message.

Traceback (most recent call last):
  File "<ipython-input-9-56ac45285836>", line 15, in <module>
    s = p.get_structure(pdbid, filename)
  File "C:\Anaconda\lib\site-packages\Bio\PDB\MMCIFParser.py", line 25, in get_structure
    self._build_structure(structure_id)
  File "C:\Anaconda\lib\site-packages\Bio\PDB\MMCIFParser.py", line 118, in _build_structure
    structure_builder.init_chain(current_chain_id)
  File "C:\Anaconda\lib\site-packages\Bio\PDB\StructureBuilder.py", line 82, in init_chain
    if self.model.has_id(chain_id):
AttributeError: 'StructureBuilder' object has no attribute 'model'

**Comment by ofajardo** [comment]
I run again my program against 4600 cif files, and it is interesting that I get again the old different error, vs when I run only for those 2 files where I get your error. Anyway, hopefully by fixing them the errors are gone anyway.

**Comment by JoaoRodrigues** [comment]
Would you mind posting the code you are running, or at least the relevant
part? Maybe it's a problem of how the parser is being setup.

A qui, 3/03/2016, 08:49, ofajardo <notifications@github.com> escreveu:

**Comment by ofajardo** [comment]
Hi, 

It was actually quiet easy to reproduce the problem with a minimum code. The first requisite is to parse 4GTR before 4q96. The second requisite is that the line where the parser object is built (p=MMCIFParser() ) is done once and outside the loop, if I move this line inside the loop, meaning I am building the object fresh on every loop, the problem disappears. Is it incorrect to build the object only once? my thinking was that if I am going to loop though a bunch of structures, it would save time building it only once.

p = MMCIFParser()
structs = ["4gtr","4q96","4q94"]
for pdbid in structs:
    filename = os.path.join(cif_folder,pdbid+".cif")
    print filename
    s = p.get_structure(pdbid,filename)

**Comment by peterjc** [comment]
Good work. This does sound like an initialisation / state problem when a parser class is reused. I agree it would seem natural and efficient to be able to do this. If it isn't really possible, then a clear error would be much better.

P.S. You can use the triple ` markup on GitHub to surround a block of code (I edited your comment)

**Comment by JoaoRodrigues** [comment]
@ofajardo Went over this again, the error is not reproducible anymore. I'm pushing a few changes to the main branch, would you mind (if and after these are accepted) going over your structures again to see if any errors persist?

**Comment by ofajardo** [comment]
Hi, 

I will very happy to test everything again. So, now in theory I should get the same results as when parsis the pdb format file counterpart?

**Comment by JoaoRodrigues** [comment]
Yes, in principle, you should get the same number of chains, which was what
was differing before.

Thanks a lot!

A qua, 6/04/2016, 04:29, ofajardo <notifications@github.com> escreveu:

**Comment by ofajardo** [comment]
Hi, 

Yes, indeed the error described in this ticket is not there anymore. Thanks a lot for fixing it and great job! Now I get another problem, which I will describe in a separate ticket.

Otto
